// Copyright 2018 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

import "google/protobuf/any.proto";

package paxos;

service Paxos {
  rpc Echo (EProposal) returns (EResponse) {}
  rpc Prepare (Proposal) returns (Response) {}
  rpc Accept (Proposal) returns (Response) {}
  rpc Decide (Proposal) returns (Response) {}
  rpc Ping (EmptyMessage) returns (EmptyMessage) {}
}

message EProposal {
	int32  seq = 1;
	string value = 2;
}

message EResponse {
  int32  seq = 1;
	string value = 2;
}

message MetaData {
	int me = 1;
	int done = 2;
}

message Proposal {
	int                 proposed_num = 1;
	int                 seq = 2;
	google.protobuf.Any value = 3;
	MetaData            meta = 4;
  string              type = 5;
}

message Response {
  bool                approved = 1;
  int                 number = 2;
  google.protobuf.Any value = 3;
  MetaData            meta = 4;
  string              type = 5;
}

message EmptyMessage {}
